### ISSUE TITLE
favicon.ico from custom theme is not applied

### ISSUE THREAD

**Issue by mmccarn** [owncloud/core#23758]
### Steps to reproduce
1. Edit config.php to specify   'theme' => 'example',

### Expected behaviour
favicon will be served from "/themes/example/core/img/favicon.ico":
`<link rel="shortcut icon" href="/themes/example/core/img/favicon.ico">`

### Actual behaviour
favicon.ico is served from /core/img:
`<link rel="shortcut icon" href="/core/img/favicon.ico">`

### Additional Troubleshooting steps / comments
If I replace 'favicon.ico' with 'favicon.png' in core/templates/layout.user.php and refresh my browser, the correct file is served.
`<link rel="shortcut icon" href="/themes/example/core/img/favicon.png">`

I have confirmed that favicon.ico does exist in the example theme, and is accessible from the web:
https://my-owncloud-server/themes/example/core/img/favicon.ico

It seems that this line determines the correct theme path for for .png files but not for .ico files:
`print_unescaped(image_path($_['appid'], 'favicon.ico')); /* IE11+ supports png */ ?>">`

### Server configuration
**Operating system**:
Ubuntu 14.04

**Web server:**
Apache/2.4.7 (Ubuntu)

**Database:**
 5.5.47-0ubuntu0.14.04.1 (Ubuntu)

**PHP version:**
PHP 5.5.9-1ubuntu4.14

**ownCloud version:** (see ownCloud admin page)
  - version: 9.0.0.19

**Updated from an older ownCloud or fresh install:**
- updated from 8.2

**Where did you install ownCloud from:**
Repositories

**Signing status (ownCloud 9.0 and above):**

```
Technical information
=====================
The following list covers which files have failed the integrity check. Please read
the previous linked documentation to learn more about the errors and how to fix
them.

Results
=======
- core
	- INVALID_HASH
		- core/img/favicon.ico
		- lib/private/route/cachingrouter.php
	- EXTRA_FILE
		- lib/private/route/cachingrouter.php.org
- files_external
	- INVALID_HASH
		- lib/smb.php
- user_external
	- INVALID_HASH
		- lib/imap.php

Raw output
==========
Array
(
    [core] => Array
        (
            [INVALID_HASH] => Array
                (
                    [core/img/favicon.ico] => Array
                        (
                            [expected] => 9894ca649dbbf6bcda4d77965744c2695ad071570bbedb9509db9747747f6f139b6ed9483360c607feff8eeb105b323fb8089b2b9d067d77b91530d81b09b898
                            [current] => 17bc4bb185e4b9c077d6d4ac40876bf81c4e9a87badeb05c074e1ba5a15df5d6e2d129793dbc45640b87cabe42ba4a15c8131cc7ac5392d09b4c56ecf6efa061
                        )

                    [lib/private/route/cachingrouter.php] => Array
                        (
                            [expected] => d246e73235b17d79ec37f9d4fa2110c3b68cc4584893fe0be5c983f9dc2a65ae6e2098cbd4df0e45b090296be9f583dfe8ca5ed47fb880277e44279c80f127e7
                            [current] => 2468462b71bbcdf4e7b5a161ab4d8ef0330e06275ea611064130f8f0e4881072041399d87a34c5e930daf586d50cfca89247d95c38137f03dec45287558e9982
                        )

                )

            [EXTRA_FILE] => Array
                (
                    [lib/private/route/cachingrouter.php.org] => Array
                        (
                            [expected] => 
                            [current] => 
                        )

                )

        )

    [files_external] => Array
        (
            [INVALID_HASH] => Array
                (
                    [lib/smb.php] => Array
                        (
                            [expected] => 823db849be4a8e785848d760fddd79c008eeebbf2390dc821b391a634d8e49db61a0e3f641eafe71dea2686d86189d8d80ff2f467f674e27990916c22d225ecd
                            [current] => b29a2d69106e703f1eb92244fa0b6d0adc3048ca82497b9b215b3ad1d0f45fb0a644420689fbee9a6b8d37d2f352c234e3a613dc690e396d5383757305ba46f4
                        )

                )

        )

    [user_external] => Array
        (
            [INVALID_HASH] => Array
                (
                    [lib/imap.php] => Array
                        (
                            [expected] => d4ebccdc4a2f2a97456c666ef6f6608ed631ceaee3f605e270df88be08a1c3e2c3b6944f07a271612bc6fa79b708c0d9a2ee54b310e68255c415f46904665efa
                            [current] => 8c8bd0fb53e94920c1b773dfdafb7fea7d223b9600ebcb328d72cbf36f232b5bc19ac6f2f13e81e226569cfc3e26ea7a73eb00da32bc306b216204a140c4d152
                        )

                )

        )

)
```

Explanations of signature failures:

- files_external/lib/smb.php customized to remove '@domain.tld' from users before mounting external share
- user_external/lib/imap.php customized to select IMAP server by '@domain.tld' and add new users group 'domain.tld' automatically
- core/img/favicon.ico customized for my installation
- lib/private/route/cachingrouter.php [patched](https://github.com/owncloud/core/pull/23214/commits/c3c491689b2243b88fe88f4ffb1f8274575d2029) to fix a problem with the activity app when using memcached

**List of activated apps:**

```
sudo -u www-data ./occ app:list
Enabled:
  - activity: 2.2.1
  - comments: 0.2
  - dav: 0.1.5
  - federatedfilesharing: 0.1.0
  - federation: 0.0.4
  - files: 1.4.4
  - files_external: 0.5.2
  - files_pdfviewer: 0.8
  - files_sharing: 0.9.1
  - files_videoplayer: 0.9.8
  - firstrunwizard: 1.1
  - gallery: 14.5.0
  - notifications: 0.2.3
  - provisioning_api: 0.4.1
  - systemtags: 0.2
  - templateeditor: 0.1
  - updatenotification: 0.1.0
  - user_external: 0.4
Disabled:
  - encryption
  - external
  - files_texteditor
  - files_trashbin
  - files_versions
  - user_ldap

```

**The content of config/config.php:**

```
sudo -u www-data ./occ config:list system
{
    "system": {
        "instanceid": "ocojcz4c8jtx",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "192.168.1.7",
            "REMOVED"
        ],
        "datadirectory": "\/vcloud\/data",
        "overwrite.cli.url": "",
        "dbtype": "mysql",
        "version": "9.0.0.19",
        "dbname": "owncloud",
        "dbhost": "localhost",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "theme": "example",
        "user_backends": [
            {
                "class": "OC_User_IMAP",
                "arguments": [
                    "{mail06.aicr.org:993\/imap\/ssl\/novalidate-cert\/notls\/readonly}"
                ]
            },
            {
                "class": "OC_User_IMAP",
                "arguments": [
                    "{office.mmsionline.us:993\/imap\/ssl\/novalidate-cert\/notls\/readonly}"
                ]
            }
        ],
        "maintenance": false,
        "forcessl": true,
        "forceSSLforSubdomains": true,
        "mail_smtpmode": "smtp",
        "mail_from_address": "vcloud-noreply",
        "mail_domain": "aicr.org",
        "mail_smtphost": "mail06.aicr.org",
        "loglevel": 3,
        "logtimezone": "America\/New_York",
        "log_authfailip": true,
        "log_rotate_size": 104857600,
        "memcache.local": "\\OC\\Memcache\\Memcached",
        "memcache.distributed": "\\OC\\Memcache\\Memcached",
        "memcached_servers": [
            [
                "localhost",
                11211
            ]
        ],
        "cache_path": "",
        "knowledgebaseenabled": true,
        "enable_avatars": true,
        "allow_user_to_change_display_name": true,
        "preview_libreoffice_path": "\/usr\/bin\/libreoffice",
        "enabledPreviewProviders": [
            "OC\\Preview\\PNG",
            "OC\\Preview\\JPEG",
            "OC\\Preview\\GIF",
            "OC\\Preview\\BMP",
            "OC\\Preview\\XBitmap",
            "OC\\Preview\\MP3",
            "OC\\Preview\\TXT",
            "OC\\Preview\\MarkDown",
            "OC\\Preview\\MSOfficeDoc",
            "OC\\Preview\\PDF",
            "OC\\Preview\\SVG",
            "OC\\Preview\\Photoshop"
        ],
        "asset-pipeline.enabled": true,
        "apps_paths": [
            {
                "path": "\/var\/www\/owncloud\/apps",
                "url": "\/apps",
                "writable": true
            },
            {
                "path": "\/var\/www\/owncloud\/apps-aicr",
                "url": "\/apps-aicr",
                "writable": false
            }
        ],
        "xframe_restriction": true,
        "mail_smtpsecure": "ssl",
        "mail_smtpauthtype": "PLAIN",
        "mail_smtpport": "465",
        "mail_smtpauth": 1,
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***",
        "updatechecker": false,
        "activity_expire_days": 3652
    }
}

```

**Are you using external storage, if yes which one:** local/smb/sftp/...
SFTP, FTP, SMB/CIFS

**Are you using encryption:** yes/no
NO

**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...
IMAP

### Client configuration
**Browser:**
- Firefox 45.0.1
- Internet Explorer 11.162.10586.0
- Microsoft Edge 25.10586 / Microsoft EdgeHTML 13.10586

**Operating system:**
Windows 10 Version 10.0.10586

#### Web server error log
```
[Sun Apr 03 08:58:17.954598 2016] [mpm_prefork:notice] [pid 3522] AH00163: Apache/2.4.7 (Ubuntu) PHP/5.5.9-1ubuntu4.14 OpenSSL/1.0.1f configured -- resuming normal operations
[Sun Apr 03 08:58:17.954645 2016] [core:notice] [pid 3522] AH00094: Command line: '/usr/sbin/apache2'

```

#### ownCloud log (data/owncloud.log)
not useful

#### Browser log
not useful

**Comment by LukasReschke** [comment]
cc @owncloud/designers

**Comment by mmccarn** [comment]
The issue seems to be in the image selection logic in /lib/private/urlgenerator.php.

I've proposed a patch here:
https://github.com/owncloud/core/compare/master...mmccarn:patch-2?diff=unified&name=patch-2

**Comment by jancborchardt** [comment]
@mmccarn can you open a pull request with that? cc @schiesbn @MorrisJobke

**Comment by mmccarn** [comment]
@jancborchardt I've never done that -- do I just click the 'pull request' button on that patch?

**Comment by MorrisJobke** [comment]
> @jancborchardt I've never done that -- do I just click the 'pull request' button on that patch?

Yes :)

**Comment by tcitworld** [comment]
This should have been closed in https://github.com/owncloud/core/pull/24766, but it still doesn't work with git master.

**Comment by DeepDiver1975** [comment]
Backport requested in the pr. Closing this issue.
Fixes with #24766
